### PR TITLE
Fix optimizer dtype

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

Bug fixes 
### PR changes

APIs
### Describe

The dtype of Optimizer should be inferred by given parameters, which will be used in learning rate and other vars in optimizer.

For example,
```
class MyLayer(paddle.nn.Layer):
    def __init__(self, dtype):
        super(MyLayer, self).__init__()
        self._w = self.create_parameter([2, 3], dtype='float64')
        self._b = self.create_parameter([2, 3], dtype='float64')

    def forward(self, x):
        return x * self._w + self._b
            model = MyLayer(dtype)

x = paddle.rand([10, 2, 3], dtype=dtype)
loss = model(x)
adam = paddle.optimizer.Adam(parameters=model.parameters())
loss.backward()
adam.step()
```

### CODE REVIEW

**Review by chenwhql — APPROVED** [review]
LGTM
